### PR TITLE
Adding wallet network functionality, restructuring settings page

### PR DESCRIPTION
**Pull request by hariria**
## Motivation

This PR's primary focus was on creating a switch network feature

- [x] You are now able to change networks on the wallet. (ie. from Devnet to Localhost). 
- [x] Also restructured settings page to include a section on credentials, network, and sign out
- [x] Some more refactoring

## Test Plan

https://user-images.githubusercontent.com/31301117/170852812-b0f375c4-2b8f-4f47-9ddf-2fab17aac772.mov

### CODE REVIEW

File: ecosystem/web-wallet/src/core/components/NetworkBody.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    title: 'Localhost',
+    value: LOCAL_NODE_URL,
+  },
+];
+
+export default function NetworkBody() {
+  const {
+    aptosNetwork,
+  } = useWalletState();
+  const { data: testnetIsLive } = useQuery('getTestnetStatus', getLocalhostIsLive, { refetchInterval: 1000 });
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/components/NetworkBody.tsx:41]: Why is it `testnetIsLive` when the query is `getLocalhostIsLive`?

File: ecosystem/web-wallet/src/core/components/NetworkListItem.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+} from '@chakra-ui/react';
+import { LOCAL_NODE_URL } from 'core/constants';
+import {
+  AptosNetwork,
+  networkUriMap,
+} from 'core/utils/network';
+import React from 'react';
+
+export interface SettingsListItemProps {
+  title?: 'Mainnet' | 'Testnet' | 'Devnet' | 'Localhost';
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/components/NetworkListItem.tsx:21]: Seems like this is repeated from NetworkPreference, it should use a shared type (e.g. the `networkUriMap`).

File: ecosystem/web-wallet/src/core/components/NetworkListItem.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  value: AptosNetwork;
+}
+
+export const secondaryBgColor = {
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/components/NetworkListItem.tsx:25]: Why are they exported?

File: ecosystem/web-wallet/src/core/components/NetworkListItem.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+          isLoading ? (
+            <>
+              <Text fontWeight={600}>
+                {value ? networkUriMap[value] : undefined}
+              </Text>
+              <Text fontWeight={400}>
+                {value}
+              </Text>
+              {
+                (isDisabled && value === LOCAL_NODE_URL) ? (
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/components/NetworkListItem.tsx:74]: This logic is repeated from earlier, it should be extracted to a variable.

File: ecosystem/web-wallet/src/core/components/NetworkListItem.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  return (
+    <Box as="label">
+      <input disabled={isDisabled && (value === LOCAL_NODE_URL)} {...input} />
+      <Box
+        {...checkbox}
+        cursor="pointer"
+        borderRadius="md"
+        bgColor={secondaryBgColor[colorMode]}
+        _checked={{
+          bg: 'teal.600',
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/components/NetworkListItem.tsx:52]: For consistency, move these to a variable like the other colors (even if the color should be the same in light/dark mode).

File: ecosystem/web-wallet/src/core/components/SettingsListItem.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  icon,
+  path,
+  title,
+}: SettingsListItemProps) {
+  const navigate = useNavigate();
+  const { colorMode } = useColorMode();
+  const { signOut } = useWalletState();
+
+  const gridOnClick = () => {
+    if (title === 'Sign out') {
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/components/SettingsListItem.tsx:51]: This looks fragile. Can it be changed to (for example) `type === 'sign_out'`?

File: ecosystem/web-wallet/src/core/mutations/network.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    // switching to local testnet
+    if (newAptosNetwork === LOCAL_NODE_URL && !testnetIsLive) {
+      return;
+    }
+    const accountExists = await getAccountExists({
+      address: aptosAccount?.address().hex(),
+      nodeUrl: newAptosNetwork,
+    });
+    if (!accountExists && aptosAccount) {
+      const faucetClient = new FaucetClient(newAptosNetwork, newFaucetNetwork);
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/mutations/network.tsx:32]: It is surprising that switching the network will create an account. I think this should be separate from the logic of switching the network.

File: ecosystem/web-wallet/src/core/queries/network.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import axios from 'axios';
+import { LOCAL_FAUCET_URL, LOCAL_NODE_URL } from 'core/constants';
+
+export const getLocalhostIsLive = async () => {
+  try {
+    const localNode = await axios.get(LOCAL_NODE_URL);
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/queries/network.ts:6]: Can these requests be made in parallel?

File: ecosystem/web-wallet/src/core/utils/network.ts [hunk 9]
```diff
@@ -1,13 +1,38 @@
...
     WALLET_STATE_NETWORK_LOCAL_STORAGE_KEY,
-  ) as AptosNetwork | null);
+  ) as AptosNetwork | null) || DEVNET_NODE_URL;
+}
+
+export function getFaucetNetworkFromDevnet(aptosNetwork: AptosNetwork): FaucetNetwork {
+  switch (aptosNetwork) {
+    case DEVNET_NODE_URL: return faucetUriMap.DEVNET_NODE_URL;
+    case LOCAL_NODE_URL: return faucetUriMap.LOCAL_NODE_URL;
+    default: return faucetUriMap.DEVNET_NODE_URL;
```
Comment (zacharydenton) [inline ecosystem/web-wallet/src/core/utils/network.ts:36]: The default case should be `assertNever` for exhaustiveness checking.

**Review by zacharydenton — APPROVED** [review]

File: ecosystem/web-wallet/src/core/components/CredentialsBody.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            <DrawerOverlay />
+            <DrawerContent>
+              <DrawerHeader borderBottomWidth="1px" px={4}>
+                Credentials
+              </DrawerHeader>
+              <DrawerBody px={4}>
+                <VStack mt={2} spacing={4} pb={8}>
+                  <CredentialHeaderAndBody
+                    header="Private key"
+                    body={privateKeyHex}
```
[the remark this answers is not in the document]
Comment (hariria) [reply ecosystem/web-wallet/src/core/components/CredentialsBody.tsx:54]: In wallet layout it checks if an aptos account exists, if not, it signs them out 

```jsx
  useEffect(() => {
    if (!aptosAccount) {
      navigate('/');
    }
  }, []);
```

File: ecosystem/web-wallet/src/core/components/NetworkBody.tsx (hunk 1 above)
Comment (hariria) [reply ecosystem/web-wallet/src/core/components/NetworkBody.tsx:41]: It's local testnet, i will fix the wording

File: ecosystem/web-wallet/src/core/components/NetworkBody.tsx [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import { DEVNET_NODE_URL, LOCAL_NODE_URL } from 'core/constants';
+import useWalletState from 'core/hooks/useWalletState';
+import { useQuery } from 'react-query';
+import { getLocalhostIsLive } from 'core/queries/network';
+import useSwitchNetwork from 'core/mutations/network';
+import NetworkListItem from './NetworkListItem';
+
+interface NetworkPreference {
+  description?: string;
+  title?: 'Mainnet' | 'Testnet' | 'Devnet' | 'Localhost';
```
[the remark this answers is not in the document]
Comment (hariria) [reply ecosystem/web-wallet/src/core/components/NetworkBody.tsx:22]: fixed

File: ecosystem/web-wallet/src/core/components/NetworkListItem.tsx (hunk 3 above)
Comment (hariria) [reply ecosystem/web-wallet/src/core/components/NetworkListItem.tsx:25]: Nowhere :)

File: ecosystem/web-wallet/src/core/mutations/network.tsx (hunk 7 above)
Comment (hariria) [reply ecosystem/web-wallet/src/core/mutations/network.tsx:32]: @zacharydenton people will be confused if they switch networks and there's no account available for them to transfer money to or they can't use the faucet there

File: ecosystem/web-wallet/src/core/queries/network.ts (hunk 8 above)
Comment (hariria) [reply ecosystem/web-wallet/src/core/queries/network.ts:6]: Yeah i can, but its querying localhost which shouldn't have any latency issues...

File: ecosystem/web-wallet/src/pages/Settings.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    </SimpleGrid>
+  );
+}
+
+function Account() {
+  const { aptosAccount } = useWalletState();
+
+  const privateKeyObject = aptosAccount?.toPrivateKeyObject();
+  const address = privateKeyObject?.address;
+  const explorerAddress = `https://explorer.devnet.aptos.dev/account/${address}`;
```
[the remark this answers is not in the document]
Comment (hariria) [reply ecosystem/web-wallet/src/pages/Settings.tsx:55]: `https://explorer.devnet.aptos.dev/` doesn't change based on the network you're on. It's the same regardless of if you're on devnet, testnet, or local

**Pull request merged**